src/components/IOUConfirmationList.js

```diff
@@ -328,84 +321,174 @@ class IOUConfirmationList extends Component {
     * @param {Object} option
     */
     toggleOption(option) {
-        // Return early if selected option is currently logged in user.
+        // Return early if selected option is currently logged-in user.
         if (option.login === this.props.session.email) {
             return;
         }
 
-        this.setState((prevState) => {
-            const newParticipants = _.map(prevState.participants, (participant) => {
-                if (participant.login === option.login) {
-                    return {...option, selected: !option.selected};
-                }
-                return participant;
-            });
-            return {participants: newParticipants};
-        });
+        this.setState(
+            (prevState) => {
+                const newParticipants = _.map(prevState.participants, (participant) => {
+                    if (participant.login === option.login) {
+                        return {...option, selected: !option.selected};
+                    }
+                    return participant;
+                });
+                const newSelectedParticipants = _.where(newParticipants, {selected: true});
+                const newSections = this.getSections(newParticipants);
+
+                return {
+                    sections: newSections,
+                    participants: newParticipants,
+                    selectedParticipants: newSelectedParticipants,
+                };
+            },
+            () => { this.allOptions = OptionsListUtils.flattenSections(this.state.sections); },
+        );
+    }
+
+    /**
+     * Scrolls to the focused index within the SectionList
+     *
+     * @param {Number} index
+     */
+    scrollToIndex(index) {
+        if (!this.props.canModifyParticipants) {
+            return;
+        }
+
+        const option = this.allOptions[index];
+        if (!this.list || !option) {
+            return;
+        }
+
+        const {index: itemIndex, sectionIndex} = option;
+
+        // Note: react-native's SectionList automatically strips out any empty sections.
+        // So we need to reduce the sectionIndex to remove any empty sections in front of the one we're trying to scroll to.
+        // Otherwise, it will cause an index-out-of-bounds error and crash the app.
```

Comment:
**question:** Why would an `index-out-of-bounds` error crash the app? Does implementing [`onScrollToIndexFailed`](https://reactnative.dev/docs/0.66/virtualizedlist#onscrolltoindexfailed) not protect us or is that used for something else?

Reply:
> Does implementing [onScrollToIndexFailed](https://reactnative.dev/docs/0.66/virtualizedlist#onscrolltoindexfailed) not protect us

Yeah, I'm not 100% sure but I think this is a bug in `SectionList`. We [do have `onScrollToIndexFailed` implemented](https://github.com/Expensify/App/blob/495d3db81c5bc8a84868078ffb7a06868b4ab795/src/components/OptionsList/BaseOptionsList.js#L163) and I was still seeing a crash.

Reply:
> Why would an index-out-of-bounds error crash the app? 

It's just an uncaught JS exception in SectionList that lands us in the error boundary.

Reply:
Ok, thanks. I'm still a little lost on what this logic is doing though...

If every valid option has a `sectionIndex` and a `itemIndex` why do we need to "decrement" here? Is this something we can do in the `flattenSections` step and provide indices that won't throw errors?

Reply:
So this is the only comment I haven't resolved with code changes. I tried just removing this code and the crash is easily reproducible.

> If every valid option has a sectionIndex and a itemIndex why do we need to "decrement" here?

So from what I have seen in my testing, if we have sections like this:

```json
{
  {
    sectionIndex: 0,
    data: [
      {itemIndex: 0},
      {itemIndex: 1},
    ],
  },
  {
    sectionIndex: 1,
    data: [],
  },
  {
    sectionIndex: 2,
    data: [
      {itemIndex: 0},
    ],
  },
}
```

Then we do `this.list.scrollToLocation({sectionIndex: 2, itemIndex: 0})` will cause an `indexOutOfBoundsError` and a crash. I'm pretty sure this is a RN bug.

I can work on a minimal reproduction, open an issue, and _maybe_ even submit a PR to fix it. But that will take some time and I would prefer not to HOLD this PR on that process.